### ISSUE TITLE
How to bypass RubyGems prompts in CI - bundler's executable "bundle" conflicts with ...

### ISSUE THREAD

**Issue by MSP-Greg** [rubygems/rubygems#5245]
Using WIndows Ruby 3.1 on Actions.  When trying to perform:
```
gem install bundler -v "~> 2"
```

The following error is generated:

```
  ERROR:  Error installing bundler:
  	"bundle" from bundler conflicts with C:/hostedtoolcache/windows/Ruby/3.1.0/x64/bin/bundle
```

When running locally, RubyGems prompts with:

```
bundler's executable "bundle" conflicts with C:/rubies/Ruby31-x64/bin/bundle
Overwrite the executable? [yN]
```

What's the recommended way to bypass the prompt in CI?

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by MSP-Greg** [comment]
JFYI, `gem update --system` overwrites the bin files with no prompt.

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

Happy New Year!  Don't know if you (or anyone else) have time for this issue, but please see https://github.com/ruby/setup-ruby/pull/224#issuecomment-1004979258.

Most of this is due to the setup-ruby action installing Bundler 2.3.4.  Maybe the best thing to do is update RubyGems for Ruby 2.3 thru 2.5, and just update Bundler for Ruby 2.6 and later.  I.think.  We may have a fix for the issue with Windows Ruby 3.1.0, which was the original issue...

**Comment by MSP-Greg** [comment]
Windows - It depends on one's setup, but Windows users may often have two bash shells they can use, one from Git, and one from MSYS2.  So, Windows Rubies have two bin files for stdlibs/gems that run from the command line, which includes gem and bundle.  One file has a .cmd or .bat extension, the other (the bash one) has no extension.

Updating bundler creates a bash script with the shebang line coded to Ruby's actual path.  The shipped bash script's shebang uses `#!/usr/bin/env ruby`, since the installation directory is chosen by the user.

At present I can't think of any way to fix the issue.  Not sure whether to close...

**Comment by eregon** [comment]
@MSP-Greg Could you show the `bin/bundler` you have locally on Windows (either through RubyInstaller or mswin/mingw-loco)?
My theory is on Windows Ruby changes those files at build/install times and hence we get this error.
But I'd like to verify the file is indeed modified on Windows.

**Comment by MSP-Greg** [comment]
@eregon Thanks, I reviewed things again, and the issue isn't the above.  The following prefixes the normal bash bin files:
```
:""||{ ""=> %q<-*- ruby -*-
@"%~dp0ruby" -x "%~f0" %*
@exit /b %ERRORLEVEL%
};{ #
bindir="${0%/*}" #
exec "$bindir/ruby" "-x" "$0" "$@" #
>,
}
```

I think it was added to allow the bash bin files and the Windows bin files to be the same.  It's added in Ruby's [`tool/rbinstall.rb`](https://github.com/ruby/ruby/blob/533bc771705381c39ca0ac97d116f5c2ed452237/tool/rbinstall.rb#L485-L490).  I think that's where the change should occur?  Lars and I can patch around it, but fixing the source is probably better.  JFYI, I'm still locked out at https://bugs.ruby-lang.org/ due to the wonderful TFA...
